app/src/main/java/com/pvp/app/ui/screen/goals/GoalViewModel.kt

```diff
@@ -0,0 +1,218 @@
+        }
+    }
+
+    fun changeMonthly() {
+        _state.update { currentState ->
+            currentState.copy(
+                monthly = !currentState.monthly
+            )
+        }
+
+        filterGoals()
+    }
+
+    private fun filterGoals(
+        goals: List<Goal>,
+        start: LocalDate,
+        end: LocalDate,
+        monthly: Boolean
+    ): List<Goal> {
+        return goals.filter { goal ->
+            goal.startDate.isBefore(end.plusDays(1))
+                    && goal.endDate.isAfter(start.minusDays(1))
+                    && goal.monthly == monthly
+        }
+    }
+
+    private fun filterGoals() {
+        val (start, end) = when (state.value.monthly) {
+            true -> {
+                Pair(
+                    state.value.monthStartDate,
+                    state.value.monthEndDate
+                )
+            }
+
+            false -> {
+                Pair(
+                    state.value.weekStartDate,
+                    state.value.weekEndDate
+                )
+            }
+        }
+
+        _state.update { currentState ->
+            currentState.copy(
+                currentGoals = filterGoals(
+                    goals = currentState.allGoals,
+                    start = start,
+                    end = end,
+                    monthly = currentState.monthly
+                )
+            )
+        }
+    }
+
+    private suspend fun getMonthSteps(): Long {
+        val end = Instant.now()
+
+        val start = LocalDate
+            .now()
+            .minusDays(30)
+            .atStartOfDay(ZoneId.systemDefault())
+            .toInstant()
+
+        return healthConnectService.aggregateSteps(
+            start = start,
+            end = end
+        )
+    }
+
+    fun next() {
+        _state.update { currentState ->
+            currentState.copy(
+                monthStartDate = currentState.monthStartDate.plusMonths(1),
+                monthEndDate = currentState.monthEndDate.plusMonths(1),
+                weekStartDate = currentState.weekStartDate.plusDays(7),
+                weekEndDate = currentState.weekEndDate.plusDays(7)
+            )
+        }
+
+        filterGoals()
+    }
+
+    fun previous() {
+        _state.update { currentState ->
+            currentState.copy(
+                monthStartDate = currentState.monthStartDate.minusMonths(1),
+                monthEndDate = currentState.monthEndDate.minusMonths(1),
+                weekStartDate = currentState.weekStartDate.minusDays(7),
+                weekEndDate = currentState.weekEndDate.minusDays(7)
+            )
+        }
+
+        filterGoals()
+    }
+}
+
+data class GoalState(
+    val allGoals: List<Goal> = listOf(),
+    val currentGoals: List<Goal> = listOf(),
+    var monthly: Boolean = false,
+    var monthStartDate: LocalDate = LocalDate.now(),
+    var monthEndDate: LocalDate = LocalDate.now(),
+    val monthSteps: Long = 0,
+    val user: User? = null,
```

Comment:
Could become not-null if we know for certain that user is not going to be null in goal screens (passed authentication). For initial value pass `User()`.

Code could become simpler (at some places) with this change.